src/certificates.py

```diff
@@ -0,0 +1,267 @@
+                    content_commitment=False,
+                    key_encipherment=False,
+                    data_encipherment=False,
+                    key_agreement=False,
+                    key_cert_sign=True,
+                    crl_sign=True,
+                    encipher_only=False,
+                    decipher_only=False,
+                ),
+            )
+        )
+
+    existing_oids = {ext.oid for ext in cert_extensions_list}
+    for extension in csr.extensions:
+        if extension.oid == ExtensionOID.SUBJECT_ALTERNATIVE_NAME:
+            continue
+        if extension.oid in existing_oids:
+            logger.warning("Extension %s is managed by the TLS provider, ignoring.", extension.oid)
+            continue
+        cert_extensions_list.append(extension)
+
+    return cert_extensions_list
+
+
+def generate_certificate(
+    csr: str,
+    ca: str,
+    ca_key: str,
+    validity: int = 365,
+    alt_names: Optional[List[str]] = None,
+    is_ca: bool = False,
+) -> str:
+    """Generate a TLS certificate based on a CSR.
+
+    Args:
+        csr (str): CSR
+        ca (str): CA Certificate
+        ca_key (str): CA private key
+        validity (int): Certificate validity (in days)
+        alt_names (list): List of alt names to put on cert - prefer putting SANs in CSR
+        is_ca (bool): Whether the certificate is a CA certificate
+
+    Returns:
+        str: Certificate
+    """
+    csr_object = x509.load_pem_x509_csr(csr.encode())
+    subject = csr_object.subject
+    ca_pem = x509.load_pem_x509_certificate(ca.encode())
+    issuer = ca_pem.issuer
+    private_key = serialization.load_pem_private_key(ca_key.encode(), password=None)
+
+    certificate_builder = (
+        x509.CertificateBuilder()
+        .subject_name(subject)
+        .issuer_name(issuer)
+        .public_key(csr_object.public_key())
+        .serial_number(x509.random_serial_number())
+        .not_valid_before(datetime.now(timezone.utc))
+        .not_valid_after(datetime.now(timezone.utc) + timedelta(days=validity))
+    )
+    extensions = get_certificate_extensions(
+        authority_key_identifier=ca_pem.extensions.get_extension_for_class(
+            x509.SubjectKeyIdentifier
+        ).value.key_identifier,
+        csr=csr_object,
+        alt_names=alt_names,
+        is_ca=is_ca,
+    )
+    for extension in extensions:
+        try:
+            certificate_builder = certificate_builder.add_extension(
+                extval=extension.value,
+                critical=extension.critical,
+            )
+        except ValueError as e:
+            logger.warning("Failed to add extension %s: %s", extension.oid, e)
+
+    cert = certificate_builder.sign(private_key, hashes.SHA256())  # type: ignore[arg-type]
+    return cert.public_bytes(serialization.Encoding.PEM).decode().strip()
+
+
+def generate_ca(
+    private_key: str,
+    subject: str,
+    validity: int = 365,
+    country: str = "US",
```

Comment:
Are the defaults necessary? If I'm not mistaken the country is not mandatory and the validity is as important as the subject for the caller to specify.

Reply:
`Country` should probably come from user config if we want to keep it at all.

Reply:
While I do agree with those comments, here I simply took what was in the lib v3 and moved it to self-signed certificates. Making the country a Juju config makes sense but I would see this as a separate task.

Reply:
Sure that makes sense. Let's do it in a separate task.

Reply:
Ticket created.